text/0605-eks-rewrite.md

```diff
@@ -0,0 +1,476 @@
+  handlerRole,
+});
+```
+
+After
+
+```
+const kubectlProvider = eks.KubectlProvider.fromKubectlProviderArn(this, 'KubectlProvider', {
+  functionArn: // Required. ARN of the original kubectl function
+});
+```
+
+Following parameters are removed:
+
+- kubectlRoleArn
+- handlerRole
+
+`fromKubectlProviderAttributes()` is renamed to
+`fromKubectlProviderFunctionArn()`.
+
+Reason: when the KubectlProvider was created in another stack, the lambda
+execution role already has permissions to access the cluster.
+
+Besides that, optional KubectlProvider specific properties are moved into
+KubectlProviderOptions to better organize properties.
+
+KubectlProviderOptions Definition
+
+```
+export interface KubectlProviderOptions {
+  readonly role?: iam.IRole;
+  readonly awscliLayer?: lambda.ILayerVersion;
+  readonly kubectlLayer?: lambda.ILayerVersion;
+  readonly memory?: Size;
+  readonly environment?: { [key: string]: string };
+  /**
+   * Which subnets should the provider functions be placed in.
+   */
+  readonly vpcSubnets?: ec2.SubnetSelection;
+}
+```
+
+Usage - Before
+
+```
+new eks.Cluster(this, 'MyCluster', {
+  version: eks.KubernetesVersion.V1_31,
+  kubectlMemory: Size.gigabytes(4),
+  kubectlLayer: new KubectlV31Layer(this, 'kubectl'),
+  kubectlEnvironment: {
+    'http_proxy': 'http://proxy.myproxy.com',
+  },
+  kubectlRole: iam.Role.fromRoleArn(
+    this,
+    'MyRole',
+    'arn:aws:iam::123456789012:role/lambda-role'
+  );
+});
+```
+
+Usage - After
+
+```
+new eks.Cluster(this, 'MyCluster', {
+  version: eks.KubernetesVersion.V1_31,
+  kubectlLayer: new KubectlV31Layer(this, 'kubectl'),
+  kubectlProviderOptions: {
+    memory: Size.gigabytes(4),
+    environment: {
+      'http_proxy': 'http://proxy.myproxy.com',
+    },
+    role: iam.Role.fromRoleArn(
+      this,
+      'MyRole',
+      'arn:aws:iam::123456789012:role/lambda-role'
+    );
+});
+```
+
+## Migration Path
+
+Note: We can't guarantee it's a safe migration.
+
+Due to the fact that switching from a custom resource
+(Custom::AWSCDK-EKS-Cluster) to a native L1 (AWS::EKS::Cluster) resource
+requires cluster replacement, CDK users who need to preserve their cluster will
+have to take additional actions.
+
+1. Set the authentication mode of cluster from `AuthenticationMode.CONFIG_MAP`
+   to `AuthenticationMode.API_AND_CONFIG_MAP` and deploy
+2. Set the authentication mode of cluster from
+   `AuthenticationMode.API_AND_CONFIG_MAP` to `AuthenticationMode.API` and
+   deploy
+3. Set removal policy to RETAIN on the existing cluster (and manifests) and
+   deploy.
+4. Remove cluster definition from their CDK app and deploy
+5. Add new cluster definition using the new constructs(EKSV2).
+6. Follow cdk import to import the existing cluster as the new definition.
+   1. All relevant EKS resources support import.
+   2. AWS::EKS::Cluster
+   3. AWS::EKS::FargateProfile
+   4. AWS::EKS::Nodegroup
+7. Add Manifests.
```

Comment:
Lets set a high bar for the migration experience. This will be the biggest pain-point for users.